web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/TelemetryCorrelationUtils.java

```diff
@@ -0,0 +1,382 @@
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.web.internal.correlation;
+
+import com.microsoft.applicationinsights.TelemetryConfiguration;
+import com.microsoft.applicationinsights.internal.logger.InternalLogger;
+import com.microsoft.applicationinsights.telemetry.RequestTelemetry;
+import com.microsoft.applicationinsights.web.internal.RequestTelemetryContext;
+import com.microsoft.applicationinsights.web.internal.ThreadContext;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.UUID;
+import java.util.concurrent.ThreadLocalRandom;
+
+import javax.servlet.http.HttpServletRequest;
+
+public class TelemetryCorrelationUtils {
+
+    public static final String CORRELATION_HEADER_NAME = "Request-Id";
+	public static final String CORRELATION_CONTEXT_HEADER_NAME = "Correlation-Context";
+	public static final String REQUEST_CONTEXT_HEADER_NAME = "Request-Context";
+	public static final String REQUEST_CONTEXT_HEADER_APPID_KEY = "appId";
+	public static final String REQUEST_CONTEXT_HEADER_ROLENAME_KEY = "roleName";
+	public static final int REQUESTID_MAXLENGTH = 1024; 
+
+	private TelemetryCorrelationUtils() {}
+
+	/**
+	 * Resolves correlation ID's by parsing well-known correlation headers.
+	 * @param request The servlet request.
+	 * @param requestTelemetry The request telemetry to be populated with correlation ID's.
+	 */
+	public static void resolveCorrelation(HttpServletRequest request, RequestTelemetry requestTelemetry) {
+		
+		try {
+			if (request == null) {
+				InternalLogger.INSTANCE.error("Failed to resolve correlation. request is null.");
+				return;
+			}
+	
+			if (requestTelemetry == null) {
+				InternalLogger.INSTANCE.error("Failed to resolve correlation. requestTelemetry is null.");
+				return;
+			}
+
+			String rootId = null;
+			String parentId = null;
+			String currentId = null;
+
+			String requestId = request.getHeader(CORRELATION_HEADER_NAME);
+
+			if (requestId == null || requestId.isEmpty()) {
+				// no incoming requestId, no parent.
+				rootId = generateRootId();
+				currentId = '|' + rootId + '.';
+			} else {
+				parentId = requestId;
+				rootId = extractRootId(parentId);
+				currentId = generateId(parentId);
+			}
+
+			requestTelemetry.setId(currentId);
+			requestTelemetry.getContext().getOperation().setId(rootId);
+			requestTelemetry.getContext().getOperation().setParentId(parentId);
+
+			// resolve baggages (Correlation-Context)
+			@SuppressWarnings("unchecked")
+			Enumeration<String> baggages = request.getHeaders(CORRELATION_CONTEXT_HEADER_NAME);
+
+			if (baggages == null) {
+				InternalLogger.INSTANCE.warn("Could not access header information: " + CORRELATION_CONTEXT_HEADER_NAME);
+				return;
+			}
+
+			CorrelationContext currentCorrelationContext = 
+				ThreadContext.getRequestTelemetryContext().getCorrelationContext();
+
+			while (baggages.hasMoreElements()) {
```

Comment:
can't use foreach since Enumeration is not iterable. One could use traditional for but it looks weird with enumeration... i prefer while in this case.